File: flowkit/dashboard/src/api/opencode.ts
/**
 * OpenCode client for the FlowKit dashboard.
 *
 * The browser talks to the FlowKit agent at `/api/opencode/*`. It never talks
 * to opencode.ai, and it never holds an API key: the agent owns the credential
 * and resolves which protocol each model speaks.
 *
 * Why this file looks the way it does
 * -----------------------------------
 * It used to do the opposite. It carried a live API key in the JS bundle
 * (`DEFAULT_OPENCODE_KEY`) and guessed each model's protocol from its id:
 *
 *     endpoint: isResponses ? 'responses' : 'chat/completions'
 *
 * That guess is the "models are mismatching" bug. A model whose real endpoint
 * was `responses` but whose id did not contain `muse-spark` was sent to
 * `chat/completions`, and the request failed in a way that looked like a broken
 * model rather than a routing bug. The catalogue and the protocol now come from
 * `agent/services/opencode_models.py`, which is the single source of truth.
 */

export type ReasoningEffort = 'xhigh' | 'high' | 'medium' | 'low' | 'off'

/**
 * Canonical protocol name. Note the hyphen: `chat-completions` is the *name*,
 * `/chat/completions` is the URL path. Only the server translates between them.
 */
export type EndpointType = 'responses' | 'chat-completions'

export interface ChatMessage {
  id?: string
  role: 'system' | 'user' | 'assistant'
  content: string
  timestamp?: number
  reasoningTokens?: number
  thoughtTrace?: string
  latencyMs?: number
  modelUsed?: string
  effortUsed?: ReasoningEffort
  /** Tool-activity frames collected on the agent-mode path. Absent on /chat/stream. */
  tools?: AgentToolEvent[]
}

/** A model as described by the server. Mirrors `ModelInfo.to_dict()`. */
export interface ModelOption {
  id: string
  name: string
  endpoint_type: EndpointType
  supported_reasoning: ReasoningEffort[]
  default_reasoning: ReasoningEffort
  is_free: boolean
  supports_vision: boolean
  context_window: number
  max_output_tokens: number
  /** `api` when it came from the live list, `builtin` when from the fallback. */
  source: 'api' | 'builtin'
}

/** Response of `GET /api/opencode/models`. */
export interface ModelCatalogue {
  models: ModelOption[]
  /**
   * Where the list came from. The UI shows this rather than implying the list
   * is live when the network or key was unavailable.
   */
  source: 'api' | 'builtin'
  key_configured: boolean
  error: string | null
}

/** Response of `GET /api/opencode/status`. Never contains the key itself. */
export interface OpenCodeStatus {
  configured: boolean
  base_url: string
  models_cached: number
  models_builtin: number
  source: 'api' | 'builtin'
  error: string | null
}

export interface ChatResponseResult {
  text: string
  thoughtTrace?: string
  reasoningTokens: number
  inputTokens: number
  outputTokens: number
  totalTokens: number
  latencyMs: number
  model: string
  /** Which protocol the server actually used. Surfaced so a routing bug is visible. */
  endpointType: EndpointType
  reasoningEffort: ReasoningEffort
}

/**
 * Pull a human-readable message out of an error response.
 *
 * FastAPI puts a string in `detail` for HTTPException but a list of field
 * errors for validation failures, so both shapes are handled.
 */
async function readError(res: Response): Promise<string> {
  try {
    const body = await res.json()
    const detail = body?.detail
    if (typeof detail === 'string') return detail
    if (Array.isArray(detail)) {
      return detail.map((d: any) => d?.msg ?? JSON.stringify(d)).join('; ')
    }
    if (detail) return JSON.stringify(detail)
    return `HTTP ${res.status}`
  } catch {
    return res.statusText || `HTTP ${res.status}`
  }
}

// ---------------------------------------------------------------------------
// Upstream error explanation
// ---------------------------------------------------------------------------

export interface UpstreamExplanation {
  /** Short, plain-language statement of what went wrong. */
  title: string
  /** What the person can actually do about it. */
  hint: string
  /** False when retrying the same model could still succeed. */
  switchModel: boolean
}

/**
 * Turn a provider failure into something a person can act on.
 *
 * OpenCode relays provider errors as a JSON envelope, occasionally nested
 * inside another JSON string, for example:
 *
 *   {"error":{"type":"CreditsError","message":"No payment method. Add a …"}}
 *
 * Rendered raw that is unreadable, and the natural reaction is to assume the
 * app itself is broken. Anything unrecognised returns `null` so the caller can
 * fall back to the original text instead of inventing a cause.
 */
export function explainUpstreamError(raw: string): UpstreamExplanation | null {
  const payload = unwrapErrorPayload(raw)
  if (!payload) return null

  const type = String(payload.type ?? '')
  const message = String(payload.message ?? '')

  if (type === 'CreditsError' || /no payment method/i.test(message)) {
    return {
      title: 'This model needs a paid account',
      hint:
        'Your OpenCode account has no payment method, so paid models are refused. ' +
        'Models whose id ends in `-free` still work — switch to one of those.',
      switchModel: true
    }
  }

  if (type === 'FreeUsageLimitError' || /free usage limit/i.test(message)) {
    return {
      title: 'The free quota for this model is used up',
      hint: 'Pick a different `-free` model, or wait for the quota to reset.',
      switchModel: true
    }
  }

  if (/model is unavailable/i.test(message)) {
    return {
      title: 'The provider has this model offline',
      hint: 'This is upstream, not local. Choose another model and try again.',
      switchModel: true
    }
  }

  if (type === 'authentication_error' || /invalid api key|unauthoriz/i.test(message)) {
    return {
      title: 'The OpenCode key was rejected',
      hint: 'Set OPENCODE_API_KEY in the repository .env, then restart the agent.',
      switchModel: false
    }
  }

  if (/rate.?limit/i.test(message)) {
    return {
      title: 'Rate limited by the provider',
      hint: 'Wait a moment and retry, or switch models.',
      switchModel: false
    }
  }

  return null
}

/**
 * Dig a provider error object out of the envelopes seen in practice.
 *
 * The payload may be an object, a JSON string, or a JSON string containing
 * another JSON string, so unwrap up to a few levels before giving up.
 */
function unwrapErrorPayload(raw: string): { type?: string; message?: string } | null {
  const levels: Record<string, any>[] = []
  let current: unknown = raw

  for (let depth = 0; depth < 4 && typeof current === 'string'; depth++) {
    const text = current.trim()
    if (!text.startsWith('{')) break
    try {
      current = JSON.parse(text)
    } catch {
      break
    }
    if (current && typeof current === 'object') {
      levels.push(current as Record<string, any>)
    }
  }

  for (const level of levels) {
    // `{type, error:{type, message}}` and `{error:{type, message}}` both occur.
    const inner =
      level.error && typeof level.error === 'object' ? (level.error as Record<string, any>) : level
    const type = inner.type ?? level.type
    const message = inner.message ?? level.message
    if (type || message) return { type, message }
  }

  return null
}

/** Whether the agent has an OpenCode key configured. */
export async function fetchOpenCodeStatus(): Promise<OpenCodeStatus> {
  const res = await fetch('/api/opencode/status')
  if (!res.ok) throw new Error(await readError(res))
  return res.json()
}

/**
 * Fetch the model catalogue from the agent.
 *
 * `refresh` bypasses the server-side TTL cache. The server falls back to its
 * built-in list rather than failing, so a network problem degrades the list
 * instead of breaking the page.
 */
export async function fetchOpenCodeModels(refresh = false): Promise<ModelCatalogue> {
  const res = await fetch(`/api/opencode/models${refresh ? '?refresh=true' : ''}`)
  if (!res.ok) throw new Error(await readError(res))
  return res.json()
}

/** Look a model up in a fetched catalogue. Returns undefined rather than guessing. */
export function findModel(models: ModelOption[], modelId: string): ModelOption | undefined {
  return models.find(m => m.id === modelId)
}

/**
 * Send a completion through the agent.
 *
 * No `apiKey` option: the key lives in server config and is injected server-side.
 * An unknown model or an unsupported reasoning tier is rejected by the server
 * with a specific message rather than silently falling back to another model.
 */
export async function sendOpenCodeChat(
  messages: { role: 'system' | 'user' | 'assistant'; content: string }[],
  options: {
    model: string
    reasoningEffort?: ReasoningEffort
    temperature?: number
  }
): Promise<ChatResponseResult> {
  const res = await fetch('/api/opencode/chat', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      model: options.model,
      messages,
      reasoning_effort: options.reasoningEffort,
      temperature: options.temperature ?? 0.7
    })
  })

  if (!res.ok) throw new Error(await readError(res))

  const data = await res.json()
  return {
    text: data.text ?? '',
    thoughtTrace: data.thought_trace ?? undefined,
    reasoningTokens: data.reasoning_tokens ?? 0,
    inputTokens: data.input_tokens ?? 0,
    outputTokens: data.output_tokens ?? 0,
    totalTokens: data.total_tokens ?? 0,
    latencyMs: data.latency_ms ?? 0,
    model: data.model,
    endpointType: data.endpoint_type,
    reasoningEffort: data.reasoning_effort
  }
}

export interface StreamUsage {
  reasoningTokens: number
  inputTokens: number
  outputTokens: number
  totalTokens: number
}

export interface StreamDone {
  model: string
  endpointType: EndpointType
  reasoningEffort: ReasoningEffort
  latencyMs: number
}

export interface StreamHandlers {
  onText?: (chunk: string) => void
  onReasoning?: (chunk: string) => void
  onUsage?: (usage: StreamUsage) => void
  onDone?: (info: StreamDone) => void
  /**
   * Extra frames emitted only by the agent-mode endpoint
   * (`POST /api/opencode/agent/stream`). Plain `/chat/stream` never sends
   * them, so a handler set without `onTool` simply ignores them.
   */
  onTool?: (event: AgentToolEvent) => void
}

/**
 * One tool-activity frame from the agent loop.
 *
 * The backend streams these between text deltas while it acts: web-search
 * results, sandboxed command output, memory-operation confirmations, and
 * project/queue snapshots. Every field except `type`/`tool` is optional
 * because each tool reports a different payload; unknown tools still arrive
 * here so the UI can render them generically instead of dropping them.
 */
export interface AgentToolEvent {
  type: 'tool'
  /** Tool name, e.g. `web_search`, `run_command`, `memory`, `project_snapshot`. */
  tool: string
  /** Lifecycle marker when the backend sends one (`start` / `done` / `error`). */
  status?: string
  /**
   * What the operation costs, from the server's operations catalog.
   * `spend` costs money, `destructive` is public/irreversible. Carried on the
   * frame so the card can warn *before* the result is read.
   */
  risk?: 'read' | 'write' | 'spend' | 'destructive'
  /** True when the server refused the call on risk grounds — nothing ran. */
  refused?: boolean
  /**
   * Tool arguments as sent (when the backend includes them). Optional:
   * current frames carry tool-specific fields (command/query/results/…) but
   * not raw args. The destructive-approval card therefore re-confirms via
   * the chat loop (model re-invokes with {"confirm": true}) instead of
   * replaying args — no new server route. If the backend ever echoes args
   * here, the UI can offer exact-args replay.
   */
  args?: Record<string, unknown>
  /** Human-readable label for the activity row. */
  title?: string
  /** `web_search`: the query that was run. */
  query?: string
  /** `web_search`: result cards. */
  results?: { title: string; url: string; snippet: string }[]
  /** `run_command`: the command that was executed. */
  command?: string
  /** `run_command`: captured output (capped server-side; UI caps rendering too). */
  output?: string
  exit_code?: number | null
  /** `memory`: which operation ran (`read` / `write` / `append` / …). */
  operation?: string
  /** `memory`: short confirmation detail. */
  detail?: string
  /** Snapshot tools: one-line summary. */
  summary?: string
  /** Snapshot tools / unknown tools: structured payload, rendered as JSON. */
  data?: unknown
  /** Error text when `status` is `error`. */
  message?: string
}

/** Translate one SSE frame, invoking the matching handler. */
function handleFrame(frame: string, handlers: StreamHandlers): void {
  for (const line of frame.split('\n')) {
    if (!line.startsWith('data:')) continue
    const raw = line.slice(5).trim()
    if (!raw) continue

    let event: any
    try {
      event = JSON.parse(raw)
    } catch {
      // One malformed frame is not worth aborting the whole reply over.
      continue
    }

    switch (event.type) {
      case 'text':
        handlers.onText?.(event.text ?? '')
        break
      case 'reasoning':
        handlers.onReasoning?.(event.text ?? '')
        break
      case 'usage':
        handlers.onUsage?.({
          reasoningTokens: event.reasoning_tokens ?? 0,
          inputTokens: event.input_tokens ?? 0,
          outputTokens: event.output_tokens ?? 0,
          totalTokens: event.total_tokens ?? 0
        })
        break
      case 'error':
        // Thrown rather than swallowed: a stream that failed must not be
        // rendered as if the model finished answering.
        throw new Error(event.message || 'The model stream failed.')
      case 'tool':
        // Agent-mode only. Plain /chat/stream never sends these; a handler
        // set without onTool drops them, which keeps the old path untouched.
        handlers.onTool?.(event as AgentToolEvent)
        break
      case 'done':
        handlers.onDone?.({
          model: event.model,
          endpointType: event.endpoint_type,
          reasoningEffort: event.reasoning_effort,
          latencyMs: event.latency_ms ?? 0
        })
        break
    }
  }
}

/**
 * Drain an SSE response, dispatching each frame to `handlers`.
 *
 * Shared by both stream endpoints so the agent path cannot drift from the
 * plain path: same framing, same error semantics, plus optional `tool`
 * frames on the agent endpoint.
 */
async function pumpSSE(res: Response, handlers: StreamHandlers): Promise<void> {
  if (!res.body) throw new Error('Streaming is unavailable in this browser.')

  const reader = res.body.getReader()
  const decoder = new TextDecoder()
  let buffer = ''

  try {
    for (;;) {
      const { done, value } = await reader.read()
      if (done) break
      buffer += decoder.decode(value, { stream: true })

      // SSE frames are separated by a blank line.
      for (;;) {
        const sep = buffer.indexOf('\n\n')
        if (sep === -1) break
        const frame = buffer.slice(0, sep)
        buffer = buffer.slice(sep + 2)
        handleFrame(frame, handlers)
      }
    }
  } finally {
    reader.cancel().catch(() => {})
  }
}

/**
 * Stream a completion, reporting deltas as they arrive.
 *
 * Same validation and same server-side key injection as `sendOpenCodeChat`; the
 * difference is that text is surfaced while it is produced. Resolves when the
 * stream ends, and throws on a transport failure or an `error` frame.
 */
export async function streamOpenCodeChat(
  messages: { role: 'system' | 'user' | 'assistant'; content: string }[],
  options: {
    model: string
    reasoningEffort?: ReasoningEffort
    temperature?: number
  },
  handlers: StreamHandlers,
  signal?: AbortSignal
): Promise<void> {
  const res = await fetch('/api/opencode/chat/stream', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      model: options.model,
      messages,
      reasoning_effort: options.reasoningEffort,
      temperature: options.temperature ?? 0.7
    }),
    signal
  })

  // Rejections (unknown model, bad reasoning tier, no key) come back as an
  // ordinary JSON response before any streaming begins.
  if (!res.ok) throw new Error(await readError(res))
  await pumpSSE(res, handlers)
}

/**
 * Stream a completion through the agent tool loop.
 *
 * Same request shape as `streamOpenCodeChat`, but the server runs the
 * prompt-level tool protocol around the model (web search, memory ops,
 * project/queue snapshots, sandboxed commands) and interleaves extra `tool`
 * frames, delivered to `handlers.onTool`. Text/reasoning/usage/done frames
 * are identical, so callers accumulate them exactly as before.
 *
 * Throws an `AGENT_STREAM_NOT_FOUND …` error when the backend does not have
 * the endpoint yet, so the caller can fall back to `streamOpenCodeChat`
 * rather than showing a dead toggle.
 */
export async function streamAgentChat(
  messages: { role: 'system' | 'user' | 'assistant'; content: string }[],
  options: {
    model: string
    reasoningEffort?: ReasoningEffort
    temperature?: number
  },
  handlers: StreamHandlers,
  signal?: AbortSignal
): Promise<void> {
  const res = await fetch('/api/opencode/agent/stream', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      model: options.model,
      messages,
      reasoning_effort: options.reasoningEffort,
      temperature: options.temperature ?? 0.7
    }),
    signal
  })

  if (res.status === 404) {
    throw new Error(
      `AGENT_STREAM_NOT_FOUND: POST /api/opencode/agent/stream is not on this agent build.`
    )
  }
  if (!res.ok) throw new Error(await readError(res))
  await pumpSSE(res, handlers)
}

File: flowkit/agent/api/opencode.py
"""OpenCode access for Agent Studio.

The dashboard used to talk to ``opencode.ai`` directly through a Vite dev proxy,
carrying a hardcoded API key in its JavaScript bundle and its own guess at each
model's protocol. Two problems followed: a live credential was readable by
anyone who loaded the page, and a wrong protocol guess looked like a broken
model rather than a routing bug.

So the browser now asks this router. The key stays in server configuration, the
protocol is resolved from the catalogue, and an unknown model is rejected
outright instead of being guessed at.

Both endpoints here share one validation path (:func:`_resolve_model`) so a rule
cannot hold for the buffered call and not for the streamed one.
"""

from __future__ import annotations

import json
import logging
import time
from typing import AsyncIterator, Callable, Optional

import httpx
from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field

from agent import config
from agent.services import chat_agent
from agent.services.opencode_models import (
    ModelInfo,
    build_request_payload,
    build_stream_payload,
    endpoint_path,
    get_catalogue,
    parse_chat_response,
    parse_stream_event,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/opencode", tags=["opencode"])


class ChatMessage(BaseModel):
    role: str
    content: str


class ChatRequest(BaseModel):
    model: str
    messages: list[ChatMessage] = Field(default_factory=list)
    reasoning_effort: Optional[str] = None
    temperature: float = 0.7


class AgentChatRequest(ChatRequest):
    """A chat request that may use server-side tools."""

    use_tools: bool = True
    max_rounds: int = Field(default=chat_agent.MAX_ROUNDS, ge=1, le=8)


def _key_configured() -> bool:
    return bool(config.OPENCODE_API_KEY)


def _upstream_headers() -> dict:
    return {
        "Authorization": f"Bearer {config.OPENCODE_API_KEY}",
        "Content-Type": "application/json",
        "x-opencode-session": f"flowkit-{int(time.time())}",
        "x-opencode-request": f"req-{int(time.time())}",
        "x-opencode-client": "flowkit-agent-studio/1.0",
        "User-Agent": "flowkit-agent-studio/1.0 VSCode",
    }


def _sse(payload: dict) -> str:
    """One Server-Sent Events frame."""
    return f"data: {json.dumps(payload)}\n\n"


async def _resolve_model(body: ChatRequest) -> ModelInfo:
    """Validate a request and return the catalogue entry it names.

    Raises ``HTTPException`` rather than substituting a default: an unknown
    model, or a reasoning tier the model does not offer, is a client mistake and
    saying so is more useful than quietly doing something else.
    """
    if not _key_configured():
        raise HTTPException(
            503,
            "OpenCode is not configured. Set OPENCODE_API_KEY in the environment "
            "and restart the agent.",
        )
    if not body.messages:
        raise HTTPException(422, "messages must not be empty")

    cat = get_catalogue()
    models = await cat.get(api_key=config.OPENCODE_API_KEY, base_url=config.OPENCODE_BASE_URL)
    model = next((m for m in models if m.id == body.model), None)

    if model is None:
        # Rejecting beats guessing: a silent fallback to a different model is
        # how "the models are mismatching" presents from the outside.
        raise HTTPException(
            400,
            f"Unknown model {body.model!r}. It is not in the catalogue; "
            "refresh the model list or pick a listed model.",
        )

    if body.reasoning_effort and body.reasoning_effort not in model.supported_reasoning:
        raise HTTPException(
            422,
            f"{model.id} does not support reasoning effort "
            f"{body.reasoning_effort!r}; supported: {model.supported_reasoning}",
        )

    return model


@router.get("/status")
async def status():
    """Whether a key is configured. Never returns the key itself."""
    cat = get_catalogue()
    return {
        "configured": _key_configured(),
        "base_url": config.OPENCODE_BASE_URL,
        # ``cached`` is 0 until /models has resolved once; ``builtin`` is what
        # the UI can always fall back to. Reporting both avoids the old "0
        # models" reading that looked like a broken catalogue.
        "models_cached": cat.cached_count,
        "models_builtin": cat.builtin_count,
        "source": cat.last_source,
        "error": cat.last_error,
    }


@router.get("/models")
async def list_models(refresh: bool = False):
    """The model catalogue, resolved server-side.

    Includes the canonical free models even when the live list cannot be
    fetched, so the UI always has something selectable — and reports which
    source it used, so it does not claim to be live when it is not.
    """
    cat = get_catalogue()
    if refresh:
        cat.invalidate()

    models = await cat.get(api_key=config.OPENCODE_API_KEY or None,
                           base_url=config.OPENCODE_BASE_URL)

    return {
        "models": [m.to_dict() for m in models],
        "source": cat.last_source,
        "key_configured": _key_configured(),
        "error": cat.last_error,
    }


@router.post("/chat")
async def chat(body: ChatRequest):
    """Proxy a completion, injecting the key server-side."""
    model = await _resolve_model(body)

    payload = build_request_payload(
        model_id=model.id,
        endpoint_type=model.endpoint_type,
        messages=[m.model_dump() for m in body.messages],
        reasoning_effort=body.reasoning_effort or model.default_reasoning,
        temperature=body.temperature,
    )

    url = f"{config.OPENCODE_BASE_URL.rstrip('/')}{endpoint_path(model.endpoint_type)}"

    started = time.monotonic()
    try:
        async with httpx.AsyncClient(timeout=config.OPENCODE_TIMEOUT, trust_env=False) as client:
            response = await client.post(url, headers=_upstream_headers(), json=payload)
    except httpx.TimeoutException:
        raise HTTPException(504, f"OpenCode timed out after {config.OPENCODE_TIMEOUT:.0f}s")
    except Exception as e:
        raise HTTPException(502, f"Could not reach OpenCode: {e}")

    latency_ms = int((time.monotonic() - started) * 1000)

    if response.status_code >= 400:
        logger.warning("OpenCode %s failed (%s): %s",
                       model.id, response.status_code, response.text[:300])
        raise HTTPException(response.status_code, response.text[:500])

    parsed = parse_chat_response(model.endpoint_type, response.json())

    logger.info("OpenCode %s replied in %dms (%s)", model.id, latency_ms, model.endpoint_type)

    return {
        **parsed,
        "model": model.id,
        "endpoint_type": model.endpoint_type,
        "reasoning_effort": body.reasoning_effort or model.default_reasoning,
        "latency_ms": latency_ms,
    }


@router.post("/chat/stream")
async def chat_stream(body: ChatRequest):
    """Stream a completion as Server-Sent Events.

    Same validation and same server-side key injection as ``/chat``. The
    difference is that text arrives as it is produced rather than after the
    whole reply is buffered — a long generation used to leave the UI blank,
    which reads as "broken" rather than "slow".

    Frames are the normalised deltas from
    :func:`agent.services.opencode_models.parse_stream_event`, plus a final
    ``done`` carrying the model, the protocol actually used, and the latency.
    """
    # Validation happens before the response starts, because an exception
    # raised inside a StreamingResponse body cannot set a status code.
    model = await _resolve_model(body)

    effort = body.reasoning_effort or model.default_reasoning
    payload = build_stream_payload(
        model_id=model.id,
        endpoint_type=model.endpoint_type,
        messages=[m.model_dump() for m in body.messages],
        reasoning_effort=effort,
        temperature=body.temperature,
    )
    url = f"{config.OPENCODE_BASE_URL.rstrip('/')}{endpoint_path(model.endpoint_type)}"

    async def event_stream():
        started = time.monotonic()
        try:
            async with httpx.AsyncClient(
                timeout=config.OPENCODE_TIMEOUT, trust_env=False
            ) as client:
                async with client.stream(
                    "POST", url, headers=_upstream_headers(), json=payload
                ) as response:
                    if response.status_code >= 400:
                        raw = (await response.aread()).decode("utf-8", "replace")
                        logger.warning("OpenCode %s stream failed (%s): %s",
                                       model.id, response.status_code, raw[:300])
                        yield _sse({"type": "error", "message": raw[:500]})
                        return

                    async for line in response.aiter_lines():
                        if not line.startswith("data:"):
                            continue
                        raw = line[5:].strip()
                        if raw == "[DONE]":
                            break
                        if not raw:
                            continue
                        try:
                            data = json.loads(raw)
                        except json.JSONDecodeError:
                            # One malformed frame is not worth killing the
                            # stream over; keep going and let the UI show text.
                            continue
                        delta = parse_stream_event(model.endpoint_type, data)
                        if delta:
                            yield _sse(delta)
        except httpx.TimeoutException:
            yield _sse({
                "type": "error",
                "message": f"OpenCode timed out after {config.OPENCODE_TIMEOUT:.0f}s",
            })
            return
        except Exception as e:  # noqa: BLE001 - the client must be told, not left hanging
            logger.warning("OpenCode %s stream broke: %s", model.id, e)
            yield _sse({"type": "error", "message": f"Stream broke: {e}"})
            return

        latency_ms = int((time.monotonic() - started) * 1000)
        logger.info("OpenCode %s streamed in %dms (%s)",
                    model.id, latency_ms, model.endpoint_type)
        yield _sse({
            "type": "done",
            "model": model.id,
            "endpoint_type": model.endpoint_type,
            "reasoning_effort": effort,
            "latency_ms": latency_ms,
        })

    return StreamingResponse(
        event_stream(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            # Without this an intermediary that buffers would defeat streaming.
            "X-Accel-Buffering": "no",
        },
    )


def _stream_model_call(
    model: ModelInfo, effort: str, temperature: float, url: str
) -> Callable[[list[dict]], AsyncIterator[dict]]:
    """Build the ``model_call`` the tool loop drives.

    Returns an async generator factory: given a conversation it streams one
    completion and yields the same normalised deltas ``parse_stream_event``
    already produces. Raising here surfaces to the client as an ``error`` frame
    rather than a silent empty answer.
    """

    async def model_call(messages: list[dict]):
        payload = build_stream_payload(
            model_id=model.id,
            endpoint_type=model.endpoint_type,
            messages=messages,
            reasoning_effort=effort,
            temperature=temperature,
        )
        async with httpx.AsyncClient(
            timeout=config.OPENCODE_TIMEOUT, trust_env=False
        ) as client:
            async with client.stream(
                "POST", url, headers=_upstream_headers(), json=payload
            ) as response:
                if response.status_code >= 400:
                    raw = (await response.aread()).decode("utf-8", "replace")
                    logger.warning(
                        "OpenCode %s agent stream failed (%s): %s",
                        model.id, response.status_code, raw[:300],
                    )
                    raise RuntimeError(raw[:500])

                async for line in response.aiter_lines():
                    if not line.startswith("data:"):
                        continue
                    chunk = line[5:].strip()
                    if chunk == "[DONE]":
                        break
                    if not chunk:
                        continue
                    try:
                        data = json.loads(chunk)
                    except json.JSONDecodeError:
                        continue
                    delta = parse_stream_event(model.endpoint_type, data)
                    if delta:
                        yield delta

    return model_call


@router.post("/agent/stream")
async def agent_chat_stream(body: AgentChatRequest):
    """Stream a completion with server-side tools available.

    Same validation and key injection as ``/chat/stream``; the difference is
    that the reply is scanned for tool markers, which the server executes
    itself, feeding results back for up to ``max_rounds`` before the final
    answer is streamed. This is what lets the assistant act instead of handing
    the user commands to paste.

    Adds two frame types on top of the existing vocabulary: ``tool`` (with
    ``status`` of ``running``/``ok``/``error``) and ``error``.
    """
    model = await _resolve_model(body)
    effort = body.reasoning_effort or model.default_reasoning
    url = f"{config.OPENCODE_BASE_URL.rstrip('/')}{endpoint_path(model.endpoint_type)}"

    messages = [m.model_dump() for m in body.messages]
    if body.use_tools:
        messages = chat_agent.prepare_messages(messages)

    model_call = _stream_model_call(model, effort, body.temperature, url)

    async def event_stream():
        started = time.monotonic()
        rounds = 0
        try:
            async for event in chat_agent.run_agent(
                messages, model_call, max_rounds=body.max_rounds
            ):
                if event.get("type") == "tool":
                    rounds += 1
                yield _sse(event)
        except Exception as e:  # noqa: BLE001 - the client must be told, not left hanging
            logger.warning("OpenCode %s agent loop broke: %s", model.id, e)
            yield _sse({"type": "error", "message": f"Agent loop broke: {e}"})
            return

        latency_ms = int((time.monotonic() - started) * 1000)
        logger.info(
            "OpenCode %s agent finished in %dms (%s, %d tool events)",
            model.id, latency_ms, model.endpoint_type, rounds,
        )
        yield _sse({
            "type": "done",
            "model": model.id,
            "endpoint_type": model.endpoint_type,
            "reasoning_effort": effort,
            "latency_ms": latency_ms,
            "tools_used": rounds > 0,
        })

    return StreamingResponse(
        event_stream(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )

